Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -97,31 +80,39 @@ public final class CacheAdvance<T: Codable> {
                 try writer.truncate(at: writer.offsetInFile)
 
                 // Set the offset back to the beginning of the file.
-                try writer.seek(to: 0)
+                try writer.seek(to: FileHeader.expectedEndOfHeaderInFile)
 
                 // We know the oldest message is at the beginning of the file, since we just tossed out the rest of the file.
-                try reader.seek(to: 0)
+                reader.offsetInFileOfOldestMessage = FileHeader.expectedEndOfHeaderInFile
+                try reader.seekToBeginningOfOldestMessage()
 
                 // We know we're about to overwrite the oldest message, so advance the reader to the second oldest message.
-                try reader.seekToNextMessage(shouldSeekToOldestMessageIfFound: false, cacheOverwritesOldMessages: true)
+                try reader.seekToNextMessage()
             }
 
             // Prepare the reader before writing the message.
             try prepareReaderForWriting(dataOfLength: bytesNeededToStoreMessage)
 
             // Create the marker for the offset representing the beginning of the message that will be the oldest once our write is done.
-            let offsetInFileOfOldestMessage = Data(Bytes(reader.offsetInFile))
+            let offsetInFileOfOldestMessage = UInt64(reader.offsetInFile)
+
+            // Update the offsetInFileOfOldestMessage in our header before we write the message.
+            // If the application crashes between writing the header and writing the message data, we'll have lost the messages between the previous offsetInFileOfOldestMessage and the new offsetInFileOfOldestMessage.
```

Comment:
I'll play with this idea – I do like the idea of having a test that proves this. But the way I've written this code I'm not going to be able to write a test like this without a pretty big architecture change. This code wast not designed to be hot-swapped / mocked.

Reply:
I see what you’re saying. Wonder if we can use the fact that we can access `internal` types in tests to mimic a crash, by recreating some of the logic here in a test. But probably not worth rearchitecting to test it. Something to keep in mind for a V2.0.

Reply:
Indeed.

In the interim, here's my reasoning for why this code does not corrupt a file if we were to crash between writes:

* If we crash between writing the `offsetInFileOfOldestMessage` and the message, we've still set the `offsetInFileOfOldestMessage` to a valid position. We've lost data, but we haven't corrupted our file.
* If we crash between writing the message and `offsetInFileAtEndOfNewestMessage`, our `offsetInFileAtEndOfNewestMessage` is still at a valid position – the beginning of the message we just wrote. We'll have lost the most recent message, but that's it.

If we manage to crash _while_ writing a message, or _while_ writing metadata to our header, we're likely in trouble. Though this was true in the last release as well.

Reply:
That logic makes sense to me.
You’d hope these file handle API write atomically (to avoid the case of crashing _while_ writing the message)  but not sure if that’s true.